File: hot100/动态规划/62.cc
#include <vector>
using namespace std;

int uniquePaths(int m, int n)
{
    vector<vector<int>> memo(m, vector<int>(n, -1)); // -1 表示没有访问过
    auto dfs = [&](auto &&dfs, int i, int j)
    {
        if(i == 0 || j == 0) return 1;
        int &res = memo[i][j]; // 注意这里是引用
        if (res != -1)
        { // 之前计算过
            return res;
        }
        // 这是个杨辉三角形，每个位置的路径 = 该位置左边的路径 + 该位置上边的路径
        return res = dfs(dfs, i - 1, j) + dfs(dfs, i, j - 1);
    };
    return dfs(dfs, m - 1, n - 1);
}

File: hot100/动态规划/62.py
from typing import List
from functools import cache
from dataclasses import dataclass
import math

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        @cache
        def dfs(i,j):
            if i==0 or j==0:
                return 1
            return dfs(i-1,j)+dfs(i,j-1)
        return dfs(m-1,n-1)